src/bp/nlu-core/entities/microsoft-extractor/index.ts

```diff
@@ -0,0 +1,380 @@
+import Recognizers from '@microsoft/recognizers-text-suite'
+import { NLU } from 'botpress/sdk'
+import { ensureFile, pathExists, readJSON, writeJson } from 'fs-extra'
+import _ from 'lodash'
+import lru from 'lru-cache'
+import ms from 'ms'
+import sizeof from 'object-sizeof'
+import path from 'path'
+
+import { EntityExtractionResult, SystemEntityExtractor, KeyedItem } from '../../typings'
+import { SystemEntityCacheManager } from '../entity-cache-manager'
+
+type SupportedLangs = 'zh' | 'nl' | 'en' | 'fr' | 'de' | 'it' | 'ja' | 'pt' | 'es'
+const SupportedLangsList = ['zh', 'nl', 'en', 'fr', 'de', 'it', 'ja', 'pt', 'es']
+
+interface MicrosoftParams {
+  lang: SupportedLangs
+  recognizers: any[]
+}
+
+const BATCH_SIZE = 10000
+
+// Further improvements:
+// 1- in _extractBatch, shift results ==> don't walk whole array n times (nlog(n) vs n2)
+
+export class MicrosoftEntityExtractor implements SystemEntityExtractor {
+  private _cache: SystemEntityCacheManager
+
+  constructor(private readonly logger?: NLU.Logger) {
+    this.logger = logger
+    this._cache = new SystemEntityCacheManager(
+      path.join(process.APP_DATA_PATH || '', 'cache', 'microsoft_sys_entities.json'),
+      true,
+      logger
+    )
+  }
+
+  private _langToCulture = (lang: SupportedLangs): string => {
+    switch (lang) {
+      case 'zh':
+        return Recognizers.Culture.Chinese
+      case 'nl':
+        return Recognizers.Culture.Dutch
+      case 'en':
+        return Recognizers.Culture.English
+      case 'fr':
+        return Recognizers.Culture.French
+      case 'de':
+        return Recognizers.Culture.German
+      case 'it':
+        return Recognizers.Culture.Italian
+      case 'ja':
+        return Recognizers.Culture.Japanese
+      case 'pt':
+        return Recognizers.Culture.Portuguese
+      case 'es':
+        return Recognizers.Culture.Spanish
+      default:
+        throw new Error('This language is not supported by Microsoft Recognizer')
```

Comment:
mmmmmmhhhhh

I would not throw in a private function like this witouth catching... Have you tried making this throw?

This should be a test case.

Also, this is a general guideline:

exception throwing is generally used in public methods as a second return stream. Its part of the interface of the class if you will. The excetion class is often exported so users can check in a catch block which exception occured and handle consequently.

In private methods, you can usually afford to return "null", "undefined" or a special return code. These mecanisms would be considered to "hackish" or unclear to be part of your class's API.